src/orderings.jl

```diff
@@ -78,16 +57,32 @@ Return whether the first word is less then the other one in a given WreathOrder
 """
 function lt(o::WreathOrder, p::T, q::T) where T<:AbstractWord{<:Integer}
     iprefix = lcp(p, q) + 1
-    dp, degreesp, headsp = @views head(p[iprefix:end])
-    dq, degreesq, headsq = @views head(q[iprefix:end])
 
-    if dp > dq
+    @views pp = p[iprefix:end]
+    @views qq = q[iprefix:end]
+
+    if isone(pp)
+        isone(qq) ? (return false) : (return true)
+    elseif isone(qq)
         return false
-    else
-        for i in 1:dp
-            @inbounds (degreesp[i] == degreesq[i]) || return (degreesp[i] < degreesq[i])
-            @inbounds (headsp[i] == headsq[i]) || return (headsp[i] < headsq[i])
+    end
+
+    max_p = maximum(pp)
+    max_q = maximum(qq)
+
+    if max_p < max_q
+        return true
+    elseif max_p == max_q
+        head_p = findall(isequal(max_p), pp)
+        head_q = findall(isequal(max_p), qq)
+        if length(head_p) < length(head_q)
+            return true
+        elseif length(head_p) == length(head_q)
+            return @views lt(o, pp[1:first(head_p) - 1], qq[1:first(head_q) - 1])
+        else
+            return false
         end
-        dp < dq ? true : false
+    else
+        return false
     end
-end
```

Comment:
I'd suggest something like:
```julia
    max_p < max_q && return true
    max_p > max_q && return false
       
    # i.e. max_p == max_q
    head_p_len = count(isequal(max_p), pp)
    head_q_len = count(isequal(max_p), qq)
    
    if head_p_len < head_q_len
        return true
    elseif head_p_len == head_q_len
        first_pp = findfirst(isequal(max_p), pp)
        first_qq = findfirst(isequal(max_p), qq)
        return @views lt(o, pp[1:first_pp - 1], qq[1:first_qq - 1])
    else
        return false
    end
```
note that `findall` allocates an array, which we discard immediately after (we use only its length)

Reply:
> Ok, so I written it in the different way. It still allocates, but it allocates half of what it used to allocate.
> 
> I also checked the resulting `RewritingSystem` obtained by our implementation for the example 5.5 and two rules are slightly different. I am not sure if this is an error in the book or we have some deeper problem. (Note that this was also the case in the former implementation of `WreathOrder`, the one using `head`).

Thanks for checking this; I don't have time to investigate it now, so either try to understand what goes wrong (probably by means of pen and paper), or open an issue, so that the potential problem is not forgotten.

Reply:
Ech, ofc. this is neat. Updated.

As to the latter: I will open an issue and look into it when I have some time.